microcosm_postgres/encryption/encryptor.py

```diff
@@ -0,0 +1,76 @@
+"""
+Implement application-layer encryption using the aws-encryption-sdk.
+
+"""
+from typing import Mapping, Sequence, Tuple, Union
+
+from aws_encryption_sdk import KMSMasterKeyProvider, encrypt, decrypt
+
+
+class KMSEncryptor:
+
+    def __init__(self, key_ids: Sequence[str]):
+        """
+        Initialize the encryptor using a sequence of master key ids (or aliases).
+
+        Encryption calls will use *all* of these master key ids as part of the encryption
+        envelope. For disaster recovery, it is considered a best practice to use multiple
+        master keys from different AWS regions.
+
+        """
+        # NB: this key provider will generate a boto session using the runtime environment
+        self.key_provider = KMSMasterKeyProvider(key_ids=key_ids)
+
+    def encrypt(self,
+                encryption_context_key: str,
+                plaintext: str) -> Tuple[bytes, Sequence[str]]:
+        """
+        Encrypt a plaintext string value.
+
+        The return value will include *both* the resulting binary ciphertext and the
+        master key ids used for encryption. In the likely case that the encryptor was initialized
+        with master key aliases, these master key ids returned will represent the unaliased key.
+
+        """
+        encryption_context = dict(
+            microcosm=encryption_context_key,
+        )
+
+        cyphertext, header = encrypt(
+            source=plaintext,
+            key_provider=self.key_provider,
+            encryption_context=encryption_context,
+        )
+
+        key_ids = [
+            encrypted_data_key.key_provider.key_info.decode("utf-8")
+            for encrypted_data_key in header.encrypted_data_keys
+        ]
+        return cyphertext, key_ids
+
+    def decrypt(self, encryption_context_key: str, ciphertext: bytes) -> str:
+        plaintext, header = decrypt(
+            source=ciphertext,
+            key_provider=self.key_provider,
+        )
+        return plaintext.decode("utf-8")
+
+
+class MultiTenantKMSEncryptor:
```

Comment:
The idea here is to use multiple sets of master key ids for different encryption contexts. This allows multi-tenancy. I'm missing:

 - A way to configure these mappings nicely with `microcosm` bindings/defaults
 - A way to handle unmapped tenants (e.g. conditional encryption)